packages/grid/x-data-grid/src/hooks/features/rows/useGridRowsMeta.ts

```diff
@@ -57,90 +57,88 @@ export const useGridRowsMeta = (
-            rowsHeightLookup.current[row.id].autoHeight = false;
+            baseRowHeight = existingBaseRowHeight;
           }
+
+          hasRowWithAutoHeight.current = true;
+          rowsHeightLookup.current[row.id].autoHeight = true;
         } else {
+          // Default back to base rowHeight if getRowHeight returns null or undefined.
+          baseRowHeight = rowHeightFromUser ?? rowHeightFromDensity;
           rowsHeightLookup.current[row.id].needsFirstMeasurement = false;
+          rowsHeightLookup.current[row.id].autoHeight = false;
         }
+      } else {
+        rowsHeightLookup.current[row.id].needsFirstMeasurement = false;
+      }
 
-        // We use an object to make simple to check if a height is already added or not
-        const initialHeights: Record<string, number> = { base: baseRowHeight };
+      // We use an object to make simple to check if a height is already added or not
+      const initialHeights: Record<string, number> = { base: baseRowHeight };
 
-        if (getRowSpacing) {
-          const indexRelativeToCurrentPage = apiRef.current.getRowIndexRelativeToVisibleRows(
-            row.id,
-          );
+      if (getRowSpacing) {
+        const indexRelativeToCurrentPage = apiRef.current.getRowIndexRelativeToVisibleRows(row.id);
 
-          const spacing = getRowSpacing({
-            ...row,
-            isFirstVisible: indexRelativeToCurrentPage === 0,
-            isLastVisible: indexRelativeToCurrentPage === currentPage.rows.length - 1,
-            indexRelativeToCurrentPage,
-          });
+        const spacing = getRowSpacing({
+          ...row,
+          isFirstVisible: indexRelativeToCurrentPage === 0,
+          isLastVisible: indexRelativeToCurrentPage === currentPage.rows.length - 1,
+          indexRelativeToCurrentPage,
+        });
 
-          initialHeights.spacingTop = spacing.top ?? 0;
-          initialHeights.spacingBottom = spacing.bottom ?? 0;
-        }
+        initialHeights.spacingTop = spacing.top ?? 0;
+        initialHeights.spacingBottom = spacing.bottom ?? 0;
+      }
 
-        const processedSizes = apiRef.current.unstable_applyPipeProcessors(
-          'rowHeight',
-          initialHeights,
-          row,
-        ) as Record<string, number>;
+      const processedSizes = apiRef.current.unstable_applyPipeProcessors(
+        'rowHeight',
+        initialHeights,
+        row,
+      ) as Record<string, number>;
 
-        rowsHeightLookup.current[row.id].sizes = processedSizes;
+      rowsHeightLookup.current[row.id].sizes = processedSizes;
 
-        const finalRowHeight = Object.values(processedSizes).reduce(
-          (acc2, value) => acc2 + value,
-          0,
-        );
-        return acc + finalRowHeight;
-      }, 0);
+      const finalRowHeight = Object.values(processedSizes).reduce((acc2, value) => acc2 + value, 0);
+      return acc + finalRowHeight;
+    }, 0);
 
+    apiRef.current.setState((state) => {
```

Comment:
I had to remove the `apiRef.current.setState`-wrapping-everything to allow a pre-processor to update the state before that the rows hydration is finished. Without this change, the pre-processor will update the state but the changes will be lost because `state` is outdated.